### PR TITLE
Add Autodesk provider to resolve #120

### LINKED ISSUE #120: Add Autodesk provider
Autodesk has a public OAuth workflow for https://developer.autodesk.com for our Forge platform and we have internal code that supports the type of middleware that this repo offers.

Before I begin the process of us setting that up to be externally contributed to here, I wanted to check a little more formally in an issue that this is still the right repo, and if there's anything specific other than the main contributing guidelines we should be aware of.

Thanks in advance!

### PR DESCRIPTION
**Pull request by hotchkj**
Generated from the Yeoman scripting, issue #120.
Tested with the sample MvcClient against https://developer.autodesk.com/ with an external account.
Added an extension that allows other claim types beyond string.

Let me know if I've missed anything from the contributing guide, thanks!

### CODE REVIEW

**Comment by hotchkj** [comment]
I'd appreciate someone giving info on the Travis build failure; it looks like a failure in an unrelated component?

File: src/AspNet.Security.OAuth.Autodesk/AutodeskAuthenticationHelper.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /// <summary>
+    /// Contains static methods that allow to extract user's information from a <see cref="JObject"/>
+    /// instance retrieved from Autodesk after a successful authentication process.
+    /// </summary>
+    public static class AutodeskAuthenticationHelper
+    {
+        /// <summary>
+        /// Key for claim containing a boolean value indicating whether the user's email has been verified
+        /// </summary>
+        public static readonly string EmailVerifiedClaim = "urn:autodesk:emailverified";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Autodesk/AutodeskAuthenticationHelper.cs]: Feel free to move these 2 constants to a new `AutodeskAuthenticationConstants` file, in a nested `Claims` class.

File: src/AspNet.Security.OAuth.Extensions/ClaimsExtensions.cs [hunk 2]
```diff
@@ -13,6 +13,11 @@ public static class ClaimsExtensions
     {
         public static ClaimsIdentity AddOptionalClaim(this ClaimsIdentity identity, string type, string value, string issuer)
         {
+            return AddOptionalClaim(identity, type, value, ClaimValueTypes.String, issuer);
+        }
+
+        public static ClaimsIdentity AddOptionalClaim(this ClaimsIdentity identity, string type, string value, string types, string issuer)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Extensions/ClaimsExtensions.cs]: `types` sounds a bit confusing. What about `valueType`?

File: src/AspNet.Security.OAuth.Autodesk/AutodeskAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public const string UserInformationEndpoint = "https://developer.api.autodesk.com/userprofile/v1/users/@me";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.Scope"/>.
+        /// </summary>
+        public static readonly IEnumerable<string> Scope = new[] { "user-profile:read" };
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Autodesk/AutodeskAuthenticationDefaults.cs]: Please consider moving `user-profile:read` to a dedicated `Scopes` nested class in `AutodeskAuthenticationConstants`.

**Comment by PinpointTownes** [comment]
Thanks for your contribution! (and for rebasing your PR... and for your patience, sorry about that :sweat:).

I'll try to merge it later today.

File: src/AspNet.Security.OAuth.Extensions/ClaimsExtensions.cs (hunk 2 above)
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Extensions/ClaimsExtensions.cs]: Or you can remove the existing `type` parameter to `name` and `types` to `types` if you prefer :+1:

**Comment by hotchkj** [comment]
Feedback should now be addressed

**Comment by PinpointTownes** [comment]
Thanks!

Did you have a chance to confirm it works as expected?
(I don't use Autodesk so it's hard to test it myself :smile:)

**Comment by PinpointTownes** [comment]
FYI, the next release should ship this week. If you want me to include your contribution in the beta3 release, please ping me to confirm it works fine.

**Comment by hotchkj** [comment]
@PinpointTownes Apologies for the delay in confirming; yes this continues to work against the actual Developer Portal. Please go ahead & merge if ready.

**Comment by PinpointTownes** [comment]
Merged: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/4d5a766b06e4b3a7c02c42ebdf2a164439ea0fec

Thanks for your contribution, @hotchkj! :clap: 

FYI, it will be included in the upcoming release (shipping this week).

**Pull request closed without merging**